### DISCUSSION TITLE
Has anyone encountered the issue of middleware not being executed during Vercel production deployment?

### DISCUSSION THREAD

**Discussion by Saszr** [amannn/next-intl#254] [Q&A]
This is the discussion I opened on Vercel and the detailed information about the problems I encountered.
https://github.com/orgs/vercel/discussions/2168

So strange 🙃

**Comment by amannn** [comment]
Hmm, interesting—I have an idea.

So what I'm seeing is this workflow:
1. Visit https://next-intl-middleware-example.vercel.app/
2. Cookie is set
3. Clear cookie locally
4. Refresh page via cmd+r
5. A 304 cache hit is returned and there is no cookie response header.
6. If I refresh via cmd+shift+r, Chrome adds the "cache-control: no-cache" request header and I get a 200 response from Vercel that includes the cookie. Interestingly, there's still the `x-vercel-cache: HIT` response header present.

So it seems like Vercel has different caching mechanisms in production.

When I open a new incognito session and request the page, I always get the cookie in the response initially. So I guess, while a bit confusing, this shouldn't be a problem practically?

**Reply by Saszr** [reply]
I guess so, Vercel production has implemented different caching strategies.
This is a serious issue in Vercel deployment and should not be ignored.

> When I open a new incognito session and request the page, I always get the cookie in the response initially. So I guess, while a bit confusing, this shouldn't be a problem practically?

The expiration time of this cookie is session-based. If you close the browser and re-enter the webpage, there will be relevant errors.

I am curious that there are very few related issues ( Including some other third-party packages. Only saw some discussions that went nowhere under next-auth. ) . It seems like everyone has not encountered this problem. The trigger for this issue is very easy.

**Reply by amannn** [reply]
> The expiration time of this cookie is session-based. If you close the browser and re-enter the webpage, there will be relevant errors.

Hmm, right—this is indeed an issue. Apart from locale matching, [the routing APIs of `next-intl`](https://next-intl-docs.vercel.app/docs/next-13/server-components#routing) will not work properly if the cookie is absent.

`next-intl` could theoretically mitigate the issue by setting a `max-age` on the response cookie, so that it persists over sessions, but ideally, this would be addressed by Vercel. Furthermore, as you're saying, `next-intl` being affected by this is only symptomatic, there might be further issues if the middleware isn't reliably executed.

I've subscribed to your inquiry at the Vercel org, hopefully this gets resolved!
